_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,323 @@
+        for (int i = 0; i < payload.length; ++i) {
+            payload[i] = (byte) (RND.nextInt(26) + 65);
+        }
+        return payload;
+    }
+
+    private boolean retriable(Exception e) {
+        if (e == null) {
+            return false;
+        } else if (e instanceof IllegalArgumentException
+                || e instanceof UnsupportedOperationException
+                || !(e instanceof RetriableException)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public void onCompletion(RecordMetadata metadata, Exception e) {
+        if (e != null) {
+            // If an exception occurred while sending the record
+            System.err.println(e.getMessage());
+
+            if (!retriable(e)) {
+                // If the exception is not retriable, print the stack trace and exit
+                e.printStackTrace();
+                System.exit(1);
+            }
+        } else {
+            // If the record was successfully sent
+            System.out.printf("Record sent to %s-%d with offset %d%n",
+                    metadata.topic(), metadata.partition(), metadata.offset());
+        }
+    }
+}
+```
+
+### Running the producer client
+
+To put this client into action, simply run the main method in the `Producer` class. 
+When running, it creates the message payloads using randomly generated byte arrays. 
+The client produces messages until it reaches the predefined message count, which is 100 messages with the `NUM_MESSAGES` constant value we specified. 
+
+> With its thread-safe design, multiple threads can share a single producer instance.
+
+### Error handling
+
+When developing a Kafka producer client, it's important to consider how you want it to handle different types of exceptions. 
+
+The error handling capabilities we introduced ensures that the producer client can recover from certain retriable errors while addressing others as non-retriable, terminating operation of the client when necessary. 
+
+Here's a breakdown of retriable and non-retriable errors that the client handles:
+
+**Non-retriable errors caught by the producer client**
+
+* `InterruptedException`: This error occurs when the current thread is interrupted while paused. 
+Interruption typically happens during producer shutdown or when stopping its operation. 
+The exception is rethrown as a `RuntimeException`, which ultimately terminates the producer.
+
+* `IllegalArgumentException`: This error is thrown when the producer receives invalid or inappropriate arguments. 
+For instance, it can be triggered if essential details like the topic are missing.
+
+* `UnsupportedOperationException`: This error is raised when an operation is not supported or when a method is not implemented as expected. 
+For instance, it can be triggered if you attempt to use an unsupported producer configuration or invoke a method that the `KafkaProducer` class does not support.
```

Comment:
How can you "invoke a method that the `KafkaProducer` class does not support"? Surely that code would not compile.

Reply:
Changed to:

> For instance, it can be triggered if you attempt to use an unsupported producer configuration, leading to a runtime exception.

Reply:
I wonder if instead you get this is the broker side does not support a feature, for example idempotency. @fvaleri do you remember what scenarios cause `UnsupportedOperationException`?